src/pad.js

```diff
@@ -0,0 +1,109 @@
+'use strict';
+
+import {Tensor} from './lib/tensor.js';
+
+/**
+ * Get mapped location from source tensor.
+ * @param {Array} location
+ * @param {Array} inputShape
+ * @param {Array} beginningPadding
+ * @param {MLPaddingMode} mode
+ * @return {Array} mapped location of source tensor
+ */
+function getMappedLocation(location, inputShape, beginningPadding, mode) {
+  const mappedLocation = location.slice();
+  const length = location.length;
+  if (mode === 'edge') {
+    for (let i = 0; i < length; i++) {
+      if (location[i] < beginningPadding[i]) {
+        mappedLocation[i] = 0;
+      } else if (location[i] >= beginningPadding[i] + inputShape[i]) {
+        mappedLocation[i] = inputShape[i] - 1;
+      } else {
+        mappedLocation[i] -= beginningPadding[i];
+      }
+    }
+  } else {
+    // reflection mode or symmetric mode
+    const offset = mode === 'symmetric' ? 1 : 0;
+    for (let i = 0; i < length; i++) {
+      if (mappedLocation[i] < beginningPadding[i]) {
+        // mappedLocation[i] =
+        //    beginningPadding[i] + (beginningPadding[i] - mappedLocation[i]) -
+        //    beginningPadding[i] - offset;
+        mappedLocation[i] = beginningPadding[i] - mappedLocation[i] - offset;
+      } else if (mappedLocation[i] >= beginningPadding[i] + inputShape[i]) {
+        // mappedLocation[i] =
+        //    beginningPadding[i] + inputShape[i] - 1 -
+        //    (mappedLocation[i] - (beginningPadding[i] + inputShape[i] -1)) -
+        //    beginningPadding[i] + offset;
+        mappedLocation[i] =
+            beginningPadding[i] + inputShape[i] * 2 - mappedLocation[i] - 2 + offset;
+      } else {
+        mappedLocation[i] -= beginningPadding[i];
+      }
+    }
+  }
+  return mappedLocation;
+}
+
+/**
+ * Update element value of destination tensor.
+ * @param {Number} index
+ * @param {Tensor} source
+ * @param {Tensor} destination
+ * @param {Array} beginningPadding
+ * @param {MLPaddingMode} mode
+ * @param {Number} value
+ * @return {Tensor}
+ */
+function updateOutput(index, source, destination, beginningPadding, mode, value) {
+  const inputShape = source.shape;
```

Comment:
[minor naming] Consider renaming the input as `input` and output as `output` (consistent like line 96), or renaming `inputShape` -> `sourceShape`.